lib/base/process.cpp

```diff
@@ -174,16 +212,65 @@ static Value ProcessSpawnImpl(struct msghdr *msgh, const Dictionary::Ptr& reques
 		sigemptyset(&mask);
 		sigprocmask(SIG_SETMASK, &mask, nullptr);
 
-		if (icinga2_execvpe(argv[0], argv, envp) < 0) {
-			char errmsg[512];
-			strcpy(errmsg, "execvpe(");
-			strncat(errmsg, argv[0], sizeof(errmsg) - strlen(errmsg) - 1);
-			strncat(errmsg, ") failed", sizeof(errmsg) - strlen(errmsg) - 1);
-			errmsg[sizeof(errmsg) - 1] = '\0';
-			perror(errmsg);
-		}
+		for (;;) {
+			if (icinga2_execvpe(argv[0], argv, envp) < 0) {
+				if (errno == E2BIG && !safeToTruncateStrings.empty()) {
+					if (safeToTruncateArgs.empty()) {
+						// Initialize safeToTruncateArgs
+						for (auto strings : {argv, envp}) {
+							for (auto it (strings); *it; ++it) {
+								auto s (*it);
+								auto len (strlen(s));
+
+								for (auto suffix : safeToTruncateStrings) {
+									if (suffix.second <= len) {
+										auto substr (s + len - suffix.second);
```

Comment:
The use of `auto` hides what's actually going on here and whether these operations are actually safe to perform in the forked process. I think this would greatly benefit from explicitly specifying the types.